Add Producers Tutorial (#877)
ManticoreGamesInc/platform-documentation: src/tutorials/producers_tutorial.md

Test the game and make sure the following works.

1. Use the **number keys** to equip items in the **Hotbar Inventory** displayed at the bottom.
2. Press ++I++ to toggle **Inventory Mode** to equip or drop items with the mouse.
3. Use the **left mouse button** to place/use items down when applicable.
4. Place a **Campfire Ring** item on the **Dirt** area and then place the **Fire Spit** item on top of it.
5. Place **Firewood** into the campfire ring and light it with the **Matches**.
6. Place the **Red Fruit** item onto the fire spit.
7. Collect the **Cooked Fruit** and wait for the fire to burn out to collect **Charcoal**.
8. Relight a new fire and place another fruit. This time do not collect until the fruit burns.
9. Use the **Hammer** item to break down and collect placed items.

<div class="mt-video" style="width:100%">
    <video autoplay muted playsinline controls loop class="center" style="width:100%">
        <source src="/img/ProducersTutorial/Producer_FruitExample.mp4" type="video/mp4" />
    </video>
</div>

!!! warning

>     While in Preview Mode, players can press ++F1++ to **reset the Player's Data**. This is useful for testing new items or features in the game. However, this should be removed before publishing a game.

Reviewer: Maybe include a little tip on where to disable it. I took a look at the framework myself just now, and could not find the setting to turn it off.

5. Drag and drop the duplicated Alien Egg Model into the **Geo** folder of the **Product - Fresh Red Fruit (Inventory Icon)**.
6. Position the duplicated Alien Egg Model to be on top of the red fruit (Basic Pepper 02).
7. Resize the duplicated Alien Egg Model to match the red fruit.
8. Delete the **Basic Pepper 02** mesh from the Geo folder.
9. Right click **Product - Fresh Red Fruit (Inventory Icon)** and select **Create New Template From This**.
10. Name the template `Alien Egg (Inventory Icon)`.
11. Delete the **Alien Egg (Inventory Icon)** template from the Hierarchy.
12. Search the Hierarchy for the `Alien Egg` script and open the Properties window.
13. Set the **IconAsset** property to **Alien Egg (Inventory Icon)**.

![!Alien Egg IconAsset](../img/ProducersTutorial/Producer_AlienEgg_IconAsset.png){: .center loading="lazy" }

#### Visualize the Icon

As long as the **Alien Egg (Inventory Icon)** matched the red fruit Icon Asset template, then it will most likely produce a clear inventory icon. However, it is good to know how the **Icon Generator** is placing the template in front of the capturing camera in case adjustments are needed.

> 1. In the Hierarchy, expand the **UI** folder and right click the **Icon Generator** and select **Desinstance This Object**.

Reviewer: Typo: Desinstance

> 1. In the Hierarchy, expand the **UI** folder and right click the **Icon Generator** and select **Desinstance This Object**.

Reviewer: The Icon Generator in the UI folder isn't a template.

Author: Are you sure? I just tested on a clean project and it seems to be a template
![template](https://user-images.githubusercontent.com/93150109/147964721-718057fa-26f9-4d76-9e8c-bb507cda926d.PNG)
.

Reviewer: For me (new project), it's not a template.

![!Burnt Egg Item](../img/ProducersTutorial/Producer_BurntEgg_Item.png){: .center loading="lazy" }

### Set the Item Properties

1. Select the **Burnt Egg** script and open the **Properties** window.
2. Set the **Name** property to `Burnt Egg`.
3. Set the **Description** property to `An egg shaped charcoal`.
4. Set the **DropTemplate** property to `Burnt Egg (Drop Icon)`.
5. Set the **IconAsset** property to `Burnt Egg (Inventory Icon)`.
6. Set the **EquipmentVisualTemplate** property to `Burnt Egg (Held)`.

![!Burnt Egg Properties](../img/ProducersTutorial/Producer_BurntEgg_Properties.png){: .center loading="lazy" }

### Add to Starting Items

The Burnt Egg item is now ready to be tested so it will need to be added as a starting item in the **Hotbar Inventory**.

1. In the Hierarchy, search for the `Starting Items` script under the **Hotbar Settings** group.
2. Right click the **Alien Egg** script and select **Duplicate**.
3. Rename the duplicated script to `Burnt Egg`.
4. Select the **Burnt Egg** script and open the Properties window.

> 5. Set the ItemId property to Burnt Egg.

Reviewer: ```suggestion
5. Set the **ItemId** property to `Burnt Egg`.
```

> 6. Set the SlotIndex property to 3.

Reviewer: ```suggestion
6. Set the **SlotIndex** property to `3`.
```

3. Right click the **Burnt Egg (Inventory Icon)** and select **Deinstance This Object**.
4. Right click the **Alien Raptor Model** in the Hierarchy and select Duplicate.
5. Drag and drop the duplicated Alien Raptor Model into the **Geo** folder of the **Burnt Egg (Inventory Icon)**.
6. Position the duplicated Alien Raptor Model to be on top of the burnt egg.
7. Resize the duplicated Alien Raptor Model to match the burnt egg.
8. Delete the **Alien Egg Model** mesh from the Geo folder.
9. Right click **Burnt Egg (Inventory Icon)** and select **Create New Template From This**.
10. Name the template `Alien Raptor (Inventory Icon)`.
11. Delete the **Alien Raptor (Inventory Icon)** template from the Hierarchy.

![!Raptor Inventory](../img/ProducersTutorial/Producer_Raptor_Inventory.png){: .center loading="lazy" }

### Create the Item Data

In the Hierarchy, open the **Global Database** folder and expand the **Items** script. Right click the **Alien Egg** script and select **Duplicate**. Rename the duplicated script to `Alien Raptor`.

![!Raptor Item](../img/ProducersTutorial/Producer_Raptor_Item.png){: .center loading="lazy" }

### Set the Item Properties

> The Alien Raptor is not be equippable so an equipment template is not needed.

Reviewer: "is not be equippable"

3. Drag and drop the **Incubator Model Off** into the **ObjectToToggle** property.
4. Activate the **HideWhenActive** property.

![!Incubator Placeable Off](../img/ProducersTutorial/Producer_Incubator_PlaceableOff.png){: .center loading="lazy" }

#### Create the Template

Right click **Incubator (Drop Icon)** and select **Create New Template From This**. Name the template `Incubator (Client)`. Delete the **Incubator (Client)** template from the Hierarchy.

![!Incubator Placeable Client](../img/ProducersTutorial/Producer_Incubator_PlaceableClient.png){: .center loading="lazy" }

### Create the Static Template

The **Static Template** will be an invisible cube to handle collision.

#### Create a Collision Cube

The **Collision Cube** needs to activate the **Can Overlap Triggers** property to avoid **overlapping Placeables**.

1. Search the **Project Content** window for the `Incubator (Drop Icon)` template.
2. Drag and drop the **Incubator (Drop Icon)** template into the scene.
3. Right click the template and select **Deinstance This Object**.
4. Search **Core Content** for a `Cube` mesh.
5. Drag and drop the **Cube** into the **Geo** group.

> 6. Transform the **Cube** match the dimensions of the **Incubator**.

Reviewer: "Transform the **Cube** match the"

2. Select the **Geo** group and rename to `Collision`.
3. Select the **Collision** group and open the **Properties** window.
4. Set **Game Collision** property to **Inherit from Parent**.
5. Set **Visibility** property to **Force Off**.
6. Set **Camera Collision** property to **Force Off**.
7. Right click **Incubator (Drop Icon)** and select **Create New Template From This**.
8. Name the template `Incubator (Static)`.
9. Delete the **Incubator (Static)** template from the Hierarchy.

![!Incubator Placeable Static Properties](../img/ProducersTutorial/Producer_Incubator_StaticProperties.png){: .center loading="lazy" }

### Create the Trigger Template

The **Trigger Template** will use triggers to check if a Placeable is going to overlap other triggers when being placed.

#### Create the Trigger

1. Search the **Project Content** window for the `Incubator (Drop Icon)` template.
2. Drag and drop the **Incubator (Drop Icon)** template into the scene.
3. Right click the template and select **Deinstance This Object**.
4. Search **Core Content** for a `Trigger`.
5. Drag and drop the **Trigger** into the **Incubator (Drop Icon)** template.

> 6. Transform the **Trigger** match the dimensions of the **Incubator Model**.

Reviewer: "Transform the **Trigger** match the"

### Test the Incubator Placeable

Preview the project and make sure to wipe the player data by pressing ++F1++.

Test the following:

- Incubators can be placed only on the Scifi area.
- Incubators cannot be placed on other Incubators.
- Placed Incubators are saved across play sessions.
- Incubators show a red light because Terminals are not placeable yet.

<div class="mt-video" style="width:100%">
    <video autoplay muted playsinline controls loop class="center" style="width:100%">
        <source src="/img/ProducersTutorial/Producer_IncubatorPlaceables.mp4" type="video/mp4" />
    </video>
</div>

## Creating a Terminal Placeable

The **Terminal Placeable** emits the Electricity Buff to nearby placed Incubators.

### Create the Client Template

The **Terminal (Drop Icon)** was scaled down to look more like a pickup, but the **Placeable Client Template** needs to be larger for the in-game size.

1. Search **Project Content** for the `Terminal (Drop Icon)` template.
2. Drag and drop the **Terminal (Drop Icon)** template into the scene.
3. Right click the template and select **Deinstance this Object**.
4. Scale up the **Terminal Model** to the in-game size.

> 5. Right click the **Terminal (Drop Icon)** template and select **Create New Template From Thi**s.

Reviewer: ```suggestion
5. Right click the **Terminal (Drop Icon)** template and select **Create New Template From This**.
```

6. Name the new template `Terminal (Client)`.

![!Terminal Placeable Client](../img/ProducersTutorial/Producer_Terminal_Client.png){: .center loading="lazy" }

### Create the Static Template

The **Static Template** is an invisible cube that handles collision.

#### Create the Collision Cube

The **Collision Cube** will use the **Terminal (Client)** template as a size reference.

1. Right click the **Terminal (Client)** template and select **Deinstance This Object**.
2. Search **Core Content** for a `Cube` mesh.
3. Drag and drop the **Cube** into the **Geo** group.

> 4. Transform the **Cube** match the dimensions of the **Terminal**.

Reviewer: "Transform the **Cube** match"

### Set the Properties

The **Terminal Placeable** is not be a Producer Base so the properties need to reflect that.

1. Inside the **Placeables** group, select the **Terminal** script and open the **Properties** window.
2. Set the **UniqueStorageId** property to `5`.
3. Set the **Client** property to `Terminal (Client)`.
4. Set the **Static** property to `Terminal (Static)`.
5. Set the **PlacementPreview** property to `Terminal (Client)`.
6. Set the **PlacementOverlapCheckTrigger** property to `Terminal (Trigger)`.
7. Set the **PlacementAreaTypes** property to `Scifi`.
8. Deactivate the **IsProducerBase** property.

![!Terminal Placeable Properties](../img/ProducersTutorial/Producer_Terminal_PlaceableProperties.png){: .center loading="lazy" }

## Testing the Placeables

Preview the game and attempt to place the Terminal and Incubator objects on the ground.

The following should work:

- Placeables can only be placed on the Scifi area.
- Incubator lights up green if within the Terminal radius.
- Incubator lights up red if not within Terminal radius.
- Placeables should be prevented from being placed on one another.
- Players should have collision with placed objects.

> - Placeables should be saved across game plays.

Reviewer: "saved across game plays" -> "saved across game sessions"

<div class="mt-video" style="width:100%">
    <video autoplay muted playsinline controls loop class="center" style="width:100%">
        <source src="/img/ProducersTutorial/Producer_TerminalPlaceables.mp4" type="video/mp4" />
    </video>
</div>

## Creating an Incubating Egg Producer

The **Incubating Egg Producer** allows an Alien Egg to be placed on the Incubator. It also handles producing drops such as the Alien Raptor and Burnt Egg.

### Examine Producer Properties

#### Producer Phases

A **Producer** uses **phases** to control the produced drops and the visuals.

These are the four phases being used for the **Incubating Egg**:

- **Placed**: When a player places the egg on the incubator it enters this phase.
- **Building**: If a placed egg is receiving the Electricity Buff, then it begins the building phase.
- **Ready**: Once the building phase has gone on for a couple seconds, it is now ready to be harvested.
- **Expired**: If not harvested when ready for a couple more seconds, then the egg expires and becomes burnt.

> 

Reviewer: Maybe switch the above list to a table with the headings "Phase" and "Description"?
